core/src/main/scala/org/apache/spark/SparkHadoopWriter.scala

```diff
@@ -105,24 +106,61 @@ class SparkHadoopWriter(@transient jobConf: JobConf)
   def commit() {
     val taCtxt = getTaskContext()
     val cmtr = getOutputCommitter()
-    if (cmtr.needsTaskCommit(taCtxt)) {
+
+    // Called after we have decided to commit
+    def performCommit(): Unit = {
       try {
         cmtr.commitTask(taCtxt)
-        logInfo (taID + ": Committed")
+        logInfo (s"$taID: Committed")
       } catch {
-        case e: IOException => {
+        case e: IOException =>
           logError("Error committing the output of task: " + taID.value, e)
           cmtr.abortTask(taCtxt)
           throw e
+      }
+    }
+
+    // First, check whether the task's output has already been committed by some other attempt
+    if (cmtr.needsTaskCommit(taCtxt)) {
+      // The task output needs to be committed, but we don't know whether some other task attempt
+      // might be racing to commit the same output partition. Therefore, coordinate with the driver
+      // in order to determine whether this attempt can commit.
+      val shouldCoordinateWithDriver: Boolean = {
+        val sparkConf = SparkEnv.get.conf
+        // We only need to coordinate with the driver if there are multiple concurrent task
+        // attempts, which should only occur if speculation is enabled
```

Comment:
yes, i think there is a possible to have multiple concurrently running task attempts for the same task when there are fetchFailed resultTasks and spark.shuffle.service.enabled = false. example: after one minute, one failed resultTask reports fetchFailed and DAGScheduler makes stage failedStage. then DAGScheduler will submit all uncompleted tasks of this result stage and some of uncompleted tasks have been submitted before. so at this time, it has multiple concurrently running task attempts for the same task. @JoshRosen @vanzin